src/mlpack/core/optimizers/fw/frank_wolfe.hpp

```diff
@@ -0,0 +1,166 @@
+    typename LinearConstrSolverType,
+    typename UpdateRuleType>
+class FrankWolfe
+{
+ public:
+  /**
+   * Construct the Frank-Wolfe optimizer with the given function and
+   * parameters. Notice that the constraint domain \f$ D \f$ is input
+   * at the initialization of linear_constr_solver, the function to be  
+   * optimized is stored in update_rule.
+   *
+   * @param linearConstrSolver Solver for linear constrained problem.
+   * @param updateRule Rule for updating solution in each iteration.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *     limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   */
+  FrankWolfe(const LinearConstrSolverType linearConstrSolver,
+             const UpdateRuleType updateRule,
+             const size_t maxIterations = 100000,
+             const double tolerance = 1e-10);
+
+  /**
+   * Optimize the given function using FrankWolfe.  The given starting
+   * point will be modified to store the finishing point of the algorithm,
+   * the final objective value is returned.
+   *
+   * FunctionType template class must provide the following functions:
+   *
+   *   double Evaluate(const arma::mat& coordinates);
+   *   void Gradient(const arma::mat& coordinates,
+   *                 arma::mat& gradient);
+   *
+   * @param function Function to be optimized.
+   * @param iterate Input with starting point, and will be modified to save 
+   *                the output optimial solution coordinates.
+   * @return Objective value at the final solution.
+   */
+  template<typename FunctionType>
+  double Optimize(FunctionType& function, arma::mat& iterate);
+
+  //! Get the linear constrained solver.
+  LinearConstrSolverType LinearConstrSolver()
+      const { return linearConstrSolver; }
+  //! Modify the linear constrained solver.
+  LinearConstrSolverType& LinearConstrSolver() { return linearConstrSolver; }
+
+  //! Get the update rule.
+  UpdateRuleType UpdateRule() const { return updateRule; }
+  //! Modify the update rule.
+  UpdateRuleType& UpdateRule() { return updateRule; }
+
+  //! Get the maximum number of iterations (0 indicates no limit).
+  size_t MaxIterations() const { return maxIterations; }
+  //! Modify the maximum number of iterations (0 indicates no limit).
+  size_t& MaxIterations() { return maxIterations; }
+
+  //! Get the tolerance for termination.
+  double Tolerance() const { return tolerance; }
+  //! Modify the tolerance for termination.
+  double& Tolerance() { return tolerance; }
+
+ private:
+  //! The solver for constrained linear problem in first step.
+  LinearConstrSolverType linearConstrSolver;
+
+  //! The rule to update, used in the second step.
+  UpdateRuleType updateRule;
+
+  //! The maximum number of allowed iterations.
+  size_t maxIterations;
+
+  //! The tolerance for termination.
+  double tolerance;
+};
+
+//! Orthogonal Matching Pursuit.
+using OMP = FrankWolfe<ConstrLpBallSolver, UpdateSpan>;
```

Comment:
Could you add a little extra documentation here please on what OMP is, what it's for, etc.?  Nothing too in-depth, just something simple that can point users in the right direction.  Users may come across this in the Doxygen documentation, kind of like this:

http://mlpack.org/docs/mlpack-git-old/doxygen/doxygen.php?doc=namespacemlpack_1_1neighbor.html#ad4e1ff28336994a46a680b750edaf814

so if there's not much information, it will only say `Orthogonal Matching Pursuit` without much further explanation for users to learn what it is about.